### PR TITLE
Parquet metadata-handling improvements

### PR DESCRIPTION
**Pull request by rjzamora**
This PR is intended to address some components of [Issue#5212](https://github.com/dask/dask/issues/5212) and [Issue#5200](https://github.com/dask/dask/issues/5200).

- I have added simple logic to drop statistics for any column with None values for min or max in any column chunk.  I decided not to drop *all* metadata when None is found, because it is posible that the bad statistics are isolated to column(s) that will not be used as the index (thus not affecting the divisions). This will hopefully help with #5212 
- I am including changes recommended by @martindurant [here](https://github.com/dask/dask/issues/5200#issuecomment-517382747) (note that I didn't make any real effort to "clean up" the code here, but it might be fine as is)

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/dataframe/io/parquet/arrow.py [hunk 1]
```diff
@@ -203,22 +203,27 @@ def read_metadata(
...
-                    s["columns"].append(d)
+                    if name not in skip_cols:
+                        column = row_group.column(i)
+                        d = {"name": name}
+                        if column.statistics:
+                            cs_min = column.statistics.min
+                            cs_max = column.statistics.max
+                            if None in [cs_min, cs_max]:
+                                skip_cols.add(name)
+                                continue
```
Comment (martindurant) [inline dask/dataframe/io/parquet/arrow.py]: Perfect. This should probably be done in fastparquet too when generating the stats values, but I suppose in principle, people might want to use them for something other than making dask divisions lists.

**Comment by martindurant** [comment]
I guess it would be good to measure the bytes size of the serialised complete graph for a dd.read_parquet like the one in the linked issue with reasonable number of columns and groups, with and without this change. That would be a decent proxy for the cost of doing that de/serialisation.

**Comment by martindurant** [comment]
Had a moment and did this, `len(cloudpickle.dumps(df))`, got 60002bytes with this change, 87668 without.

**Comment by martindurant** [comment]
A separate problem? https://stackoverflow.com/questions/57345240/problem-with-opening-parquet-files-with-fastparquet-after-updating-to-latest-ver

**Comment by rjzamora** [comment]
> A separate problem? https://stackoverflow.com/questions/57345240/problem-with-opening-parquet-files-with-fastparquet-after-updating-to-latest-ver

I'm actually having trouble reproducing this.  It doesn't really make much sense to me why the "partitioned_on" columns would need to be specified in the `columns` argument, or why the index would not be detected correctly.   Is the `@` notation in `'@timestamp'` significant? (i. e. I don't personally know if the @ has some important meaning)

**Comment by martindurant** [comment]
No, the "@" is just a character. 

I think the point is, that the _meta should include the columns due to directory-based partitioning whether or not they are specified in the columns= kwarg; but these columns do not show up in the schema of _metadata or _common_metadata (but they *should* be in the pandas JSON metadata).

The index thing I don't understand, since presumably the dataframe being written should have specified this. But certainly, a tests that looks like this example would be a useful addition.

**Comment by rjzamora** [comment]
> I think the point is, that the _meta should include the columns due to directory-based partitioning whether or not they are specified in the columns= kwarg; but these columns do not show up in the schema of _metadata or _common_metadata (but they should be in the pandas JSON metadata).

Ah, I think I'm understanding a bit (although still unable to reproduce the problem).  These columns should indeed be in the JSON metadata, but [they are excluded from `meta` right here](https://github.com/dask/dask/blob/4bd4ae3852767a712aa0835120fab7853faff701/dask/dataframe/io/parquet/core.py#L259).  During `compute()`, however,  fastparquet "knows" to include the partitioned-on columns, even if they are not specified in `columns`.

**Comment by martindurant** [comment]
Aha, exactly, those columns should include directory-derived ones, whether inferred from the filenames or from pandas metadata

**Comment by martindurant** [comment]
(the original meta should already include them, so this is simple to change)

**Comment by rjzamora** [comment]
> Aha, exactly, those columns should include directory-derived ones, whether inferred from the filenames or from pandas metadata

Just to be clear, this statement suggests that [this current assertion/test should **not** be true](https://github.com/dask/dask/blob/4bd4ae3852767a712aa0835120fab7853faff701/dask/dataframe/io/tests/test_parquet.py#L697), is that correct?  In `test_partition_on_cats_2`, we are making the assumption that the `columns` argument really does mean that all other columns should be left out of the resulting dataframe.

**Comment by martindurant** [comment]
Well, in previous versions, you could not unselect a directory-derived columns; but I suppose you are right and it is reasonable to do this, and then instead of changing how meta is made, we need the reader to exclude the inferred column (perhaps only possible after it was created anyway)

**Comment by martindurant** [comment]
Plus in that test, we didn't actually ensure that the data could be loaded, which I think would have found this issue.

**Comment by rjzamora** [comment]
> Plus in that test, we didn't actually ensure that the data could be loaded, which I think would have found this issue.

Are you saying that we aren't actually using `compute()`?

**Comment by martindurant** [comment]
> Are you saying that we aren't actually using compute()?

Right, for the df that explicitly checks this columns= case

**Comment by mrocklin** [comment]
@rjzamora @martindurant what is the status here?  It looks like this PR has gotten a bit stale.

**Comment by rjzamora** [comment]
> what is the status here? It looks like this PR has gotten a bit stale.

Initially, I was leaving this open in case we found any other problems.  At this point, we should probably just wrap this up and open a new PR if another issue surfaces.

**Comment by martindurant** [comment]
Agreed, I think this is an improvement

**Comment by TomAugspurger** [comment]
@rjzamora I merged master and pushed to your branch to resolve the CI issues.

**Comment by rjzamora** [comment]
> I merged master and pushed to your branch to resolve the CI issues.

Thanks @TomAugspurger !

**Comment by martindurant** [comment]
@rjzamora , up to you when to remove "WIP", I'll have a look then.

**Comment by martindurant** [comment]
Does this resolve https://stackoverflow.com/questions/57345240/problem-with-opening-parquet-files-with-fastparquet-after-updating-to-latest-ver ? Maybe the specific example there could become a test.

**Comment by rjzamora** [comment]
> Does this resolve https://stackoverflow.com/questions/57345240/problem-with-opening-parquet-files-with-fastparquet-after-updating-to-latest-ver ? Maybe the specific example there could become a test.

I am now remembering that one concern I had with this PR was that I was not able to reproduce [the  original error reported on stack overflow](https://stackoverflow.com/questions/57345240/problem-with-opening-parquet-files-with-fastparquet-after-updating-to-latest-ver).  So, although I believe the changes should address that problem, I am not actually "sure".

Note  that I also made the decision here that a call to `read_parquet` should indeed include "only" the columns specified by the `columns` argument (if not `None`), even if some missing columns correspond to partitions, etc.

**Comment by martindurant** [comment]
> I also made the decision here that a call to read_parquet should indeed include "only" the columns specified by the columns argument (if not None), even if some missing columns correspond to partitions, etc.

I am OK with that

**Comment by rjzamora** [comment]
> Maybe the specific example there could become a test.

I am hoping that `test_partition_on` has been properly expanded to cover this.

Note that I also added [this `compute` call](https://github.com/dask/dask/blob/c23136f9241ea428dc0cc733d81c0170305961a0/dask/dataframe/io/tests/test_parquet.py#L698) to `test_partition_on_cats_2` for the case where one of the "partitioned-on" columns is not specified in the `columns` argument.  Prior to these changes, this assertion would have failed because the pre- and post-compute columns would have disagreed (which seems to have caused the stack-overflow error).

**Comment by martindurant** [comment]
OK, good with me

**Comment by martindurant** [comment]
+1

**Comment by martindurant** [comment]
@rjzamora , just to check you were done here.

**Comment by rjzamora** [comment]
>  just to check you were done here.

Yes - we should open something new if any other problems show up.  Thanks for your help @martindurant !

**Comment by apiszcz** [comment]
I am seeing the same error as https://github.com/dask/dask/pull/5218#issuecomment-518299108
when reading from a parquet file created with append=True using fastparquet. I have not test the very latest dask

**Pull request merged**